samples/common/src/misc.rs

```diff
@@ -0,0 +1,222 @@
+    let mut client = DigitalTwinClient::connect(invehicle_digitial_twin_servuce_url)
+        .await
+        .map_err(|error| format!("{error}"))?;
+    let request = tonic::Request::new(FindByIdRequest { id: entity_id.to_string() });
+    let response = client.find_by_id(request).await.map_err(|error| error.to_string())?;
+    let response_inner = response.into_inner();
+    debug!("Received the response for the find_by_id request");
+    info!("response_payload: {:?}", response_inner.entity_access_info);
+
+    let entity_access_info = response_inner.entity_access_info.expect("Did not find the entity");
+
+    let mut matching_endpoint_info_option: Option<EndpointInfo> = None;
+    for endpoint_info in entity_access_info.endpoint_info_list {
+        // We require and endpoint that supports the protocol and supports all of the operations.
+        if endpoint_info.protocol == protocol
+            && is_subset(operations, endpoint_info.operations.as_slice())
+        {
+            matching_endpoint_info_option = Some(endpoint_info);
+            break;
+        }
+    }
+
+    if matching_endpoint_info_option.is_none() {
+        return Err("Did not find an endpoint that met our requirements".to_string());
+    }
+
+    let result = matching_endpoint_info_option.unwrap();
+
+    info!("Found a matching endpoint for entity id {entity_id} that has URI {}", result.uri);
+
+    Ok(result)
+}
+
+/// Use Chariott to discover the endpoint for the digital twin service.
+///
+/// # Arguments
+/// * `chariott_url` - Chariott's URL.
+pub async fn discover_digital_twin_service_using_chariott(
+    chariott_url: &str,
+) -> Result<Option<String>, Status> {
+    let mut client = ChariottServiceClient::connect(chariott_url.to_string())
+        .await
+        .map_err(|e| Status::internal(e.to_string()))?;
+
+    let request = Request::new(FulfillRequest {
+        namespace: CHARIOTT_NAMESPACE_FOR_IBEJI.to_string(),
+        intent: Some(IntentMessage { intent: Some(IntentEnum::Discover(DiscoverIntent {})) }),
+    });
+
+    // Get list of services at the requested namespace, if any.
+    let services: Option<Vec<discover_fulfillment::Service>> = client
+        .fulfill(request)
+        .await?
+        .into_inner()
+        .fulfillment
+        .and_then(|fulfillment_message| fulfillment_message.fulfillment)
+        .and_then(|fulfillment_enum| match fulfillment_enum {
+            FulfillmentEnum::Discover(discover) => Some(discover.services.into_iter().collect()),
+            _ => None,
+        });
+
+    // If we discovered one or more service, then return the URL for the first one that uses gRPC.
+    if services.is_some() {
+        for service in services.unwrap() {
+            if service.schema_kind == CHARIOTT_SCHEMA_KIND_FOR_GRPC {
+                return Ok(Some(service.url));
+            }
+        }
+    }
+
+    Ok(None)
```

Comment:
Consider changing this function to return `Result<String, Status>` instead. Replace this `Ok(None)` with `Err(Status::not_found("Failed to discover the in-vehicle digital twin service's URL, as it is not registered with Chariott"))`

Reply:
What is wrong with returning the Ok case as none to reflect that one was not found?  This gives the caller more options on what to do.

Reply:
Good suggestion.  Fixed.